Change Background Button

wgranger opened the pull request:
Closes #3187 

Describe your changes.
This PR adds a button on a user's homepage to change a profile background. Most of the line changes on this PR is rewriting the user home page into our standard ES6 and abstracting media actions into their own file to make them flexible. There's also some work done to make the button more accessible with focus, which was being difficult with simple `&:focus` styling.

# Review Checklist

- [x] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [x] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [X] Did you deploy a staging branch?
- [ ] Did you get any type checking errors from [Babel Typecheck](https://github.com/codemix/babel-plugin-typecheck)

https://edit-background.pfe-preview.zooniverse.org/

## Optional

- [ ] If it's a new component, is it in ES6? Is it clear of warnings from ESLint?
- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?

srallen commented on app/pages/home-for-user/index.jsx:
```diff
@@ -1,75 +1,76 @@
   componentDidMount() {
+    const backgroundBtn = document.getElementById('change-background');
+    backgroundBtn.addEventListener('focus', this.toggleFocus, true);
+    backgroundBtn.addEventListener('blur', this.toggleFocus, true);
```
Is there a reason why these event listeners are being added this way instead of directly on the element? If so, then I'd prefer using a ref to be consistent with using the virtual DOM instead of directly manipulating the DOM.

wgranger replied:
Gotcha. I can try using a ref. I'm not too familiar with refs, so this will be a good opportunity to work with them and learn something.

srallen replied:
Or just add `onFocus` and `onBlur` directly to the element. React supports both.

srallen replied:
I just noticed that the only thing happening is a style is being set by these handlers. This can be done entirely with CSS instead, actually. `:focus` is a css pseudo-class you can target. Then, adding an event listener to `blur` would be unnecessary as well as tracking it in the component state.

wgranger replied:
I wasn't able to get the `:focus` pseudo-class to work with this. I'm not sure if it's because of the input? However, I also noticed that other media drop buttons on PFE, with the `standard-button` class were not receiving focus although they were defined to have such in CSS.

srallen replied:
The styles work. Adding `&:hover, &:focus` to line 70 of the style file, then checking the label element using dev tools by right clicking and selecting `:focus`. What isn't working is the tab focus using keyboard. Adding `tabIndex="0"` to the element will make it work, but I'm not sure if that's the best approach. @eatyourgreens is there a better way?

wgranger replied:
Ahh. Good call. Just pushed up a commit using `tabIndex`. Restaged as well.

eatyourgreens replied:
`tabIndex` isn't needed on an input and shouldn't be used on a label (clicking on a label should focus the associated input). What you're trying to do here is style the adjacent text when the input has focus. One way to do this with just CSS is by wrapping the label text in a span and using sibling selectors `input:focus + span`.

wgranger replied:
That makes sense, but putting a `span` adjacent to the `input` tag will reveal the default "Choose file" of the `input` tag. I haven't found another way to hide this wording yet.

srallen commented on css/home-page-for-user.styl:
```diff
@@ -55,6 +55,31 @@ GRAY = #EFF2F5
+    cursor: pointer
+    font-weight: bold
+    padding: 0.8em
+    position: relative
```
This should be `position: absolute` and `top` and `left` instead. Right now, the button will move around depending on the screen size and can potentially overlap with other elements in the nav bar. Setting the margin is unnecessary with absolute positioning.

srallen commented on app/pages/home-for-user/index.jsx:
```diff
@@ -202,7 +212,16 @@ const HomePageForUser = React.createClass({
           )}
 
-          {!!hashQuery.project ? (
+          {screen.width > 700 && (
```
Is the intention to hide the button if the window is less than 700px? If so, then use `window.innerWidth` instead of screen. Right now, if I drag my browser to be smaller than 700px, the button is still visible.

wgranger commented:
I'm still open to styling suggestions on that `input`. For now, I reverted back to the `onBlur` and `onFocus` events. Also rebased, which was a bit tricky.

srallen commented:
@wgranger with @eatyourgreens suggestion, you can perhaps use `visibility: hidden;` on the input's css? Just hide it?

eatyourgreens commented:
Hiding an input with `visibility: hidden` or `display: none` removes it from the UI completely, so you can't tab to it with the keyboard. Hiding it offscreen with absolute positioning might work. OWD has some examples of hidden radio buttons, using a similar technique, for the categories down the lefthand side of the classifier at https://www.operationwardiary.org/#/classify. That might be helpful, but file inputs are a pain to style compared to radio buttons.

eatyourgreens commented:
@wgranger the `onFocus` and `onBlur` listeners should be attached to the file input, and the label element should not be in the documents tab order. At the moment, you can tab to the label, but you really want to tab to the file input, which will then open a file dialog when you press enter/space.
